### PR TITLE
stdenv/darwin: integrate a new CoreFoundation

### PR DESCRIPTION
**Pull request by copumpkin**
This also updates the bootstrap tool builder to LLVM 5, but not the ones we actually use for bootstrap. I'll make that change in a subsequent commit so as to provide traceable provenance of the bootstrap tools.

###### Motivation for this change

We have a horribly out of date CoreFoundation in our stdenv and this should modernize it a bit, allowing us to finally advance some other packages and also remove some workarounds.

###### Things done

[Here's an evaluation](https://hydra.nixos.org/eval/1478828?compare=1477805#tabs-now-fail) of the last state of this prior to me rebasing onto staging. I didn't feel like running a whole new eval after rebasing. Note that I also added the `CFFileDescriptor.h` stuff to support `watchman` properly since that eval, so if it complains that `watchman` is now failing, it's actually healthy again.

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/libraries/libuv/default.nix [hunk 1]
```diff
@@ -40,7 +40,7 @@ stdenv.mkDerivation rec {
         "tcp_open" "tcp_write_queue_order" "tcp_try_write" "tcp_writealot"
         "multiple_listen" "delayed_accept"
         "shutdown_close_tcp" "shutdown_eof" "shutdown_twice" "callback_stack"
-        "tty_pty"
+        "tty_pty" "condvar_5"
```
Comment (copumpkin) [inline pkgs/development/libraries/libuv/default.nix]: This was failing a bunch for me locally so I just turned it off. Seems unlikely that I broke that test given how flaky their other tests seem to be.

**Comment by copumpkin** [comment]
My plan is to merge this, trigger a staging eval, then submit a PR against staging for my LTO-enabled stdenv, merge that, then regenerate bootstrap tools make a new PR using those.

**Comment by copumpkin** [comment]
I've merged it but will be around for another week and will happily back it out if someone critical breaks as a result. I think `qt4` will break for a bit due to its use of a long-deprecated API in CF, but I opened https://github.com/NixOS/nixpkgs/issues/46719 in response to that. I'd rather just kill qt4 honestly.

I still welcome feedback on this PR!

**Comment by matthewbauer** [comment]
@copumpkin This ends up breaking building on macOS 10.11:

```
Undefined symbols for architecture x86_64:
  "_os_unfair_lock_lock", referenced from:
      __CFSuddenTerminationDisable in CFUtilities.c.o
      __CFSuddenTerminationEnable in CFUtilities.c.o
      __CFSuddenTerminationExitIfTerminationEnabled in CFUtilities.c.o
      __CFSuddenTerminationExitWhenTerminationEnabled in CFUtilities.c.o
      __CFMachPortCreateWithPort2 in CFMachPort.c.o
      ____CFMachPortCreateWithPort2_block_invoke in CFMachPort.c.o
      ___CFMachPortDeallocate in CFMachPort.c.o
      ...
  "_os_unfair_lock_unlock", referenced from:
      __CFSuddenTerminationDisable in CFUtilities.c.o
      __CFSuddenTerminationEnable in CFUtilities.c.o
      __CFSuddenTerminationExitIfTerminationEnabled in CFUtilities.c.o
      __CFSuddenTerminationExitWhenTerminationEnabled in CFUtilities.c.o
      __CFMachPortCreateWithPort2 in CFMachPort.c.o
      ____CFMachPortCreateWithPort2_block_invoke in CFMachPort.c.o
      ___CFMachPortDeallocate in CFMachPort.c.o
      ...
ld: symbol(s) not found for architecture x86_64
```

10.12 seems to be fine. Would it be okay to bump our minimum version to 10.12 everywhere?

**Comment by copumpkin** [comment]
Probably fine for now, maybe with a note that the major blocker is those locks so if someone cares enough and can put in a bit of work, we could probably keep it working.

**Comment by LnL7** [comment]
There's also an issue with wxmac (`CFURLGetFSRef`) https://hydra.nixos.org/build/82300955. It sounds familiar, but I can't remember if it's something you already knew about or fixed in another build.

**Comment by copumpkin** [comment]
I've seen it but haven't done anything about it. If absolutely necessary, we could probably manually resurrect the shitty `*FSRef` stuff by manually transplanting and adjusting the code from the old CF releases into the new swift one, but I'd really prefer not to if we can avoid it. Apple doesn't want to support it, and I certainly don't if they don't 😄

**Comment by LnL7** [comment]
Looks like I have a solution using cf-private, not ideal but better than creating and maintaining a whole bunch of patches.

**Comment by copumpkin** [comment]
Sounds good, thanks! I think I left a long note on `cf-private` asking people who use it to say what they need it for, so we can progressively rip it out as the situation changes.

**Comment by LnL7** [comment]
I added some comments to clarify why it's there, https://github.com/NixOS/nixpkgs/pull/49081.

**Pull request merged**